File: sword_to_offer/04.py
def find(matrix, number):
    found = False
    if matrix != None:
        rows = len(matrix)
        columns = len(matrix[0])
        row = 0
        column = columns - 1
        while row < rows and column >= 0:
            if matrix[row][column] == number:
                found = True
                break
            elif matrix[row][column] > number:
                column -= 1
            else:
                row += 1
    return found

def find1(matrix, number):
    found = False
    if matrix != None:
        rows = len(matrix)
        columns = len(matrix[0])
        row = rows - 1
        column = 0
        while row >= 0 and column < columns:
            if matrix[row][column] == number:
                found = True
                break
            elif matrix[row][column] > number:
                row -= 1
            else:
                column += 1
    return found

if __name__ == "__main__":
    matrix = [[1,2,8,9],[2,4,9,12],[4,7,10,13],[6,8,11,15]]
    number = 7
    print(find(matrix, number))

File: sword_to_offer/04.cpp
#include <iostream>
#include <vector>

using namespace std;

bool Find(int matrix, int rows, int columns, int number){
    bool found = false;
    if(matrix != nullptr && rows > 0 && columns > 0){
        int row = 0;
        int column = columns - 1;
        while(row < rows && column >= 0){
            if(matrix[row][column] == number){
                found = true;
                break;
            }
            else if(matrix[row][column] > number)
                column--;
            else
                row++;
        }
    }
}

int main(){
    int matrix[][4] = {{1,2,8,9},{2,4,9,12},{4,7,10,13},{6,8,11,15}};
    if(Find(matrix, 4, 4, 7))
        cout << "True" << endl;
    else
        cout << "False" << endl;
    return 0;
}
